File: src/main/java/net/frozenblock/configurableeverything/biome_placement/util/BiomeParameters.java
package net.frozenblock.configurableeverything.biome_placement.util;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.level.biome.Biome;
import net.minecraft.world.level.biome.Climate;

public record BiomeParameters(ResourceKey<Biome> biome, Climate.ParameterPoint parameters) {

	public static final Codec<BiomeParameters> CODEC = RecordCodecBuilder.create(instance ->
		instance.group(
			ResourceKey.codec(Registries.BIOME).fieldOf("biome_placement").forGetter(BiomeParameters::biome),
			Climate.ParameterPoint.CODEC.fieldOf("parameters").forGetter(BiomeParameters::parameters)
		).apply(instance, BiomeParameters::new)
	);
}

File: src/main/java/net/frozenblock/configurableeverything/biome_placement/util/BiomeParameters.kt
package net.frozenblock.configurableeverything.biome_placement.util

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import io.netty.buffer.ByteBuf
import net.frozenblock.lib.worldgen.biome.api.MutableParameterPoint
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.Identifier

data class BiomeParameters(
	@JvmField var biome: Identifier,
	@JvmField var parameters: MutableParameterPoint
) {
	companion object {
        @JvmField
		val CODEC: Codec<BiomeParameters> = RecordCodecBuilder.create { instance ->
			instance.group(
				Identifier.CODEC.fieldOf("biome").forGetter(BiomeParameters::biome),
				MutableParameterPoint.CODEC.fieldOf("parameters").forGetter(BiomeParameters::parameters)
			).apply(instance, ::BiomeParameters)
		}

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, BiomeParameters> = StreamCodec.composite(
            Identifier.STREAM_CODEC, BiomeParameters::biome,
            MutableParameterPoint.STREAM_CODEC, BiomeParameters::parameters,
            ::BiomeParameters
        )
	}
}
